### PR TITLE
Enable package consumers to specify a logger

### PR DESCRIPTION
**Pull request by rekmarks**
- Add named param `logger` to constructor
  - This object is assigned to a global variable `log`, which is used in place of `console` for all logging calls
  - By default, `logger = console`
- Validate `logger` param
  - Require `typeof 'object' === true` and the following methods:
    - `debug`
    - `error`
    - `info`
    - `log`
    - `trace`
  - Add `logger` validation tests

### CODE REVIEW

File: src/MetamaskInpageProvider.js [hunk 1]
```diff
@@ -20,19 +20,30 @@ const {
   NOOP,
 } = require('./utils')
 
+let log
```
Comment (rekmarks) [inline src/MetamaskInpageProvider.js:23]: Global because we don't want websites to mess with it.

File: src/MetamaskInpageProvider.js [hunk 2]
```diff
@@ -20,19 +20,30 @@ const {
...
     connectionStream,
-    { maxEventListeners = 100, shouldSendMetadata = true } = {},
+    {
+      logger = console,
+      maxEventListeners = 100,
+      shouldSendMetadata = true,
+    } = {},
   ) {
 
+    log = logger
```
Comment (brad-decker) [inline src/MetamaskInpageProvider.js:56]: [?]: Shouldn't we do some checking on the provided logger to make sure that it has the expected methods and signatures?

**Review by brad-decker — CHANGES_REQUESTED** [review]

File: src/MetamaskInpageProvider.js (hunk 2 above)
Comment (Gudahtt) [reply src/MetamaskInpageProvider.js:56]: Or alternatively, be more specific about the type of the `opts.logger` parameter? (i.e. specify that it must have any functions that we use)

It seems fair to expect consumers to pass in valid parameters if they're typed adequately, but it's not at the moment.

Comment (rekmarks) [reply src/MetamaskInpageProvider.js:56]: I added validation here: 98298c1

How do we feel about the required methods?

Comment (rekmarks) [reply src/MetamaskInpageProvider.js:56]: I'll improve the type, too.

Comment (rekmarks) [reply src/MetamaskInpageProvider.js:56]: `typedef` added in f68e8c3

**Comment by rekmarks** [comment]
`log` and `debug` are aliases in `loglevel`. Should we just require one or the other?
https://npmjs.com/package/loglevel

**Review by brad-decker — APPROVED** [review]
Changes LGTM 👍

**Comment by brad-decker** [comment]
We could either manually alias them if one is provided or not the other, or validate that one exists, but if we use one and the other is provided then we have an issue. Technically we only need to validate that the methods used in the library are provided

**Comment by Gudahtt** [comment]
I would consider `console` to be the API to emulate here, not `loglevel`. But I don't think it's important to validate the existence of any methods we don't actually use.

**Comment by rekmarks** [comment]
Fair points, although I decided to keep the properties (which in my mind are the most commonly used) in case we decide to use any of them in the future.

**Pull request merged**